pytensor/loop/op.py

```diff
@@ -0,0 +1,247 @@
+from typing import Optional
+
+import numpy as np
+
+from pytensor import In, Out
+from pytensor.compile import optdb, pfunc
+from pytensor.graph import Apply, FunctionGraph, Op, Type, node_rewriter
+from pytensor.graph.rewriting.basic import in2out
+from pytensor.scalar import constant
+from pytensor.tensor import NoneConst, and_, empty, scalar, set_subtensor
+from pytensor.tensor.type import DenseTensorType, TensorType
+from pytensor.tensor.type_other import NoneTypeT
+
+
+def validate_loop_update_types(update):
+    assert update.outputs[0].type.dtype == "bool"
+    for input_state, output_state in zip(update.inputs, update.outputs[1:]):
+        assert input_state.type == output_state.type
+
+
+class Loop(Op):
```

Comment:
TODO: Add mixit `HasInnerGraph` so that we can see the inner graph in debug_print